control-plane/agents/src/bin/core/volume/snapshot_operations.rs

```diff
@@ -155,9 +175,52 @@ impl ResourceLifecycleWithLifetime for OperationGuardArc<VolumeSnapshot> {
         registry: &Registry,
         _request: &Self::Destroy<'_>,
     ) -> Result<(), SvcError> {
-        self.start_destroy_by(registry, &()).await?;
-        let result = Ok(());
-        self.complete_destroy(result, registry).await
+        // Get the volume snapshot persisted info.
+        let volume_snapshot = self.lock().clone();
+
+        // Create a map to track the transactions upon deletion.
+        let mut updated_transactions: HashMap<SnapshotTxId, Vec<ReplicaSnapshot>> = HashMap::new();
+        let mut error_occurred = false;
+
+        // Iterate over current transactions and destroy each replica snapshot in a transaction,
+        // store the transactions and replica snapshots whose deletion failed.
+        for (txn_id, snap_list) in volume_snapshot.metadata().transactions() {
+            for snap in snap_list {
+                let mut cloned_snap = snap.clone();
+                if let Err(err) = ReplicaSnapshotDestroy::new(cloned_snap.clone())
```

Comment:
there's a not of cloning here.. why not just add a function on OperatorGuard<VS> which takes a cloned_snap, mutates it as needed and returns it on failure, then we're just moving it.

Reply:
Done.